File: stack/stack2.cpp
/* Stack implementation using linked lists in C++ with OOP principles */
#include <iostream>

/* Node class */
class Node
{
    /* Private data members */
public:
    int data;
    Node *next;
    /* Default public constructor */
public:
    Node()
        : data(), next(NULL)
    {
    }
};

/* Linked list class */
class LinkedList
{
    /* Private data members */
private:
    Node *head;
    /* Public constructor and member functions */
public:
    LinkedList()
        : head(NULL)
    {
    }

    // Push a new element onto the stack
    void pushElement(int data)
    {
        // Dynamic memory allocation on the heap for a new node
        Node *newNode = new Node();
        newNode->data = data;
        newNode->next = head;
        head = newNode;
    }

    // Pop the top element from the stack
    void popElement()
    {
        // Check if the stack is empty, there's nothing to pop
        if (!head)
        {
            std::cout << "The stack is empty!" << std::endl;
            return;
        }
        // Create a temp variable to store the top node
        Node *temp = head;
        // Shift the head to the next node and deallocate memory for the popped node
        head = temp->next;
        delete temp;
    }

    // Print the stack
    void printStack()
    {
        // Traverse the linked list (our stack) and print values of every node on each iteration
        Node *temp = head;
        while (temp)
        {
            std::cout << temp->data << " ";
            temp = temp->next;
        }
        std::cout << std::endl;
    }
};
int main()
{
    // Instantiate the linked list class
    LinkedList ll;
    ll.pushElement(5);
    ll.pushElement(6);
    ll.pushElement(10);
    ll.pushElement(65);
    ll.pushElement(66);
    ll.printStack();
    ll.popElement();
    ll.popElement();
    ll.popElement();
    ll.popElement();
    ll.popElement();
    ll.popElement();
    ll.printStack();
}

File: stack/stack2.py
# Stack DSA using linked lists in Python

# Node class
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# Linked list class


class LinkedList:
    def __init__(self):
        self.head = None

    # Methods

    # Push a new element onto the stack
    def push_element(self, data):
        # Create a new node
        newNode = Node(data)
        newNode.next = self.head
        self.head = newNode

    # Pop the top element from the stack
    def pop_element(self):
        # Check if the stack is empty, there's nothing to pop
        if (not self.head):
            print('The stack is empty!')
            return
        # otherwise shift the head to the next node
        temp = self.head
        self.head = temp.next
        temp = None

    # Print the stack
    def print_stack(self):
        # Traverse the linked list starting at the head
        temp = self.head
        while (temp):
            print(f'{temp.data} ', end='')
            temp = temp.next
        print()


if __name__ == '__main__':
    # Instantiate the linked list class
    ll = LinkedList()
    ll.push_element(5)
    ll.push_element(55)
    ll.push_element(65)
    ll.push_element(75)
    ll.push_element(87)
    ll.print_stack()
    ll.pop_element()
    ll.pop_element()
    ll.pop_element()
    ll.pop_element()
    ll.pop_element()
    ll.pop_element()
    ll.print_stack()
